evaluators/rank/NdcgEvaluator/__init__.py

```diff
@@ -0,0 +1,43 @@
+from math import log
+from typing import Sequence, Any
+
+from jina.executors.evaluators.rank import BaseRankingEvaluator
+
+
+class NDCGEvaluator(BaseRankingEvaluator):
+    """
+    :class:`NDCGEvaluator` evaluates normalized discounted cumulative gain for information retrieval.
+    """
+
+    @property
+    def metric(self):
+        return f'nDCG@{self.eval_at}'
+
+    def evaluate(self, actual: Sequence[Any], desired: Sequence[Any], *args, **kwargs) -> float:
+        """"
+        :param actual: the scores predicted by the search system.
+        :param desired: the expected score given by user as groundtruth.
+        :return the evaluation metric value for the request document.
+        """
+        actual_at_k = actual[:self.eval_at]
+        desired_at_k = actual[:self.eval_at]
+        if len(actual) < 2:
+            raise ValueError(f'Expecting gains with minimal length of 2, {len(actual)} received.')
+        dcg = self._compute_dcg(gains=actual_at_k)
+        idcg = self._compute_idcg(gains=desired_at_k)
+        if idcg == 0.0:
+            return 0.0
+        else:
+            return dcg/idcg
+
+    def _compute_dcg(self, gains):
+        """Compute discounted cumulative gain."""
+        ret = 0.0
+        for score, position in zip(gains[1:], range(2, len(gains) + 1)):
+            ret += score/log(position, 2)
```

Comment:
Some people use `2^score -1` instead of `score`.
https://en.wikipedia.org/wiki/Discounted_cumulative_gain

Reply:
yes, but this implementation lets them control it on their end if the want to change that and do the power in a crafter or something

Reply:
Since NDCG is going to be the most used metric, I think we can make this easy to user by exposing an argument which controls whether to use `score` or `2^score - 1`. This is actually quite subtle and our users might not notice. By default, we can go with `2^score - 1`

Reply:
it is a valid option.

Reply:
done